Implement generic contextual repository pattern for posts

oguzkocer opened the pull request:
## Summary

Implements generic contextual repository pattern for WordPress REST API caching and applies it to posts. WordPress returns different field sets based on the `context` parameter (`edit`, `view`, `embed`), so we need separate database tables for each context.

This PR establishes the generic infrastructure and implements it for posts. Future entities (comments, users, etc.) will follow the same pattern.

## Implementation

Added three context tables for posts:

- **Edit context** (36 columns) - Full field set including raw content, passwords, permalink templates
- **View context** (29 columns) - Public-facing fields without sensitive data or raw content
- **Embed context** (14 columns) - Minimal fields for embedding posts in other content

## Key Technical Decisions

**Lazy term loading**: The `PostContext` trait uses `FnOnce()` closures for optional data loading. This allows EmbedContext (which doesn't include categories/tags) to avoid unnecessary database queries by simply not calling the closure.

**RETURNING optimization**: All upsert operations use SQLite's `RETURNING rowid` clause instead of separate SELECT queries, eliminating 3 queries per upsert across the three contexts.

**Column index verification**: PRAGMA-based integration tests verify that column enum ordering matches actual SQL schema, preventing bugs from migration/enum mismatches.

## Architecture

- `context.rs` - Generic `IsContext` trait and marker types (EditContext, ViewContext, EmbedContext)
- `db_types/posts/` - Column enums and wrapper structs (types only, no logic)
- `repository/posts.rs` - Entity-specific `PostContext` trait and all database operations
- `docs/ADDING_NEW_ENTITIES.md` - Guide for adding future entity types (comments, users, etc.)

oguzkocer committed:
Add documentation for wp_mobile_cache contextual mappings

Added comprehensive documentation for implementing database mappings
for WordPress REST API contextual types in wp_mobile_cache.

Changes:
- Created CLAUDE.md with quick reference for common tasks
- Created docs/IMPLEMENTING_MAPPINGS.md with detailed step-by-step guide
- Documents the process for generating concrete types via cargo expand
- Explains field analysis, mapping implementation, and migration creation
- Provides helper function reference and common patterns

oguzkocer committed:
Implement view and embed context mappings for posts

Reorganized post mappings to support multiple WordPress API contexts
with separate implementations for edit, view, and embed.

Changes:
- Moved posts.rs to posts/ directory with separate files per context
- Created posts/edit.rs (existing edit context mapping)
- Created posts/view.rs (25 fields, no raw content or edit-only fields)
- Created posts/embed.rs (9 fields, minimal data for embedding)
- Created posts/mod.rs with shared helper for term relationship extraction
- All contexts use extract_categories_and_tags() helper to avoid duplication
- All 50 existing tests pass without modification

oguzkocer committed:
Refactor post repository to support multiple WordPress REST API contexts

Consolidate mapping layer into repository and add generic context support
for edit, view, and embed contexts. This eliminates unnecessary abstraction
layers and provides a cleaner architecture.

Architecture changes:
- Create new context module with IsContext and PostContext traits
- Add context marker types: EditContext, ViewContext, EmbedContext
- Create db_types module for database type definitions separate from logic
- Split db_types/posts into separate files per context (edit, view, embed)
- Remove mappings/posts directory entirely

Repository improvements:
- Make PostRepository generic over PostContext trait: PostRepository<C>
- Implement generic read operations via impl<C: PostContext>
- Keep context-specific write operations in impl PostRepository<EditContext>
- Move FromRowWithTerms trait into repository as private implementation detail
- Implement FromRowWithTerms for all three contexts (Edit, View, Embed)
- Add extract_categories_and_tags helper function

Benefits:
- Clean separation: db_types for structs, repository for logic
- Reduced code scatter - related functionality in one place
- Generic read operations work across all contexts
- Type-safe context handling via trait bounds
- Easier to navigate with split files per context

Changes:
- Add src/context.rs with context traits and marker types
- Add src/db_types/ module with posts submodule split by context
- Remove src/mappings/posts/ directory
- Refactor src/repository/posts.rs to be generic over context
- Update test fixtures to use PostRepository<EditContext>
- Add Default impl and private_bounds allow attribute

oguzkocer committed:
Return Ok(None) instead of error for not-found queries in wp_mobile_cache

Changed repository select methods to return `Result<Option<T>>` instead of
`Result<T>`, treating "not found" as a valid outcome rather than an error
condition. This is more idiomatic Rust and provides better clarity for callers.

Uses rusqlite's built-in `OptionalExtension::optional()` method to cleanly
convert `rusqlite::Result<T>` into `Result<Option<T>, SqliteDbError>`,
automatically handling `QueryReturnedNoRows` as `Ok(None)` while preserving
other errors.

Changes:
- Import `rusqlite::OptionalExtension` in repository/posts.rs
- Update `select_by_rowid` to return `Result<Option<DbPost>>`
- Update `select_by_post_id` to return `Result<Option<DbPost>>`
- Refactor query error handling using `let-else` pattern for clarity
- Update `delete_by_post_id` to handle new Option return type
- Update all test assertions to work with Option values
- Rename tests to reflect new behavior (returns_none vs returns_error)

oguzkocer committed:
Update IMPLEMENTING_MAPPINGS.md to reflect current architecture

The documentation was outdated and referenced the old structure where mapping
logic lived in `src/mappings/{entity}/`. Updated to reflect the current
architecture where database operations are properly separated.

Current architecture:
- `src/db_types/`: Type-only definitions (column enums, wrapper structs)
- `src/repository/`: All database interaction logic (FromRowWithTerms + upsert)
- `src/mappings/helpers.rs`: Reusable conversion utilities

This maintains proper separation of concerns where the repository pattern
handles all database operations in one place, avoiding split responsibilities.

Changes:
- Update Step 4 to reference `db_types/` instead of `mappings/`
- Add new Step 5 explaining FromRowWithTerms goes in repository
- Update all file path references throughout document
- Update checklist to reflect correct module locations
- Update file organization diagram with accurate structure
- Add architecture summary explaining the separation

oguzkocer committed:
Rename IMPLEMENTING_MAPPINGS.md to ADDING_NEW_ENTITIES.md

The old name no longer accurately describes the document's purpose. The guide
covers the complete process of adding database support for new WordPress
entity types, not just "implementing mappings."

Changes:
- Rename docs/IMPLEMENTING_MAPPINGS.md to docs/ADDING_NEW_ENTITIES.md
- Update self-reference in file organization diagram
- Update references in CLAUDE.md with more descriptive text

oguzkocer committed:
Remove private FromRowWithTerms trait in favor of PostContext method

Eliminated the private `FromRowWithTerms` trait by adding the `from_row_with_terms`
method directly to the `PostContext` trait. This removes the need for
`#[allow(private_bounds)]` and simplifies the architecture.

Benefits:
- No more private trait with unclear purpose
- No compiler warning suppression needed
- PostContext implementations are now complete in the repository module
- Clearer architecture: context trait defines both types and behavior

Changes:
- Add `from_row_with_terms` method to `PostContext` trait in context.rs
- Move `PostContext` implementations from context.rs to repository/posts.rs
- Replace `impl FromRowWithTerms for DbPost` with `impl PostContext for Context`
- Remove `where C::DbPost: FromRowWithTerms` bounds from repository methods
- Update call sites from `C::DbPost::from_row_with_terms` to `C::from_row_with_terms`
- Update return types from `Result<Self, ...>` to `Result<Self::DbPost, ...>`

oguzkocer committed:
Use lazy closure for term relationship loading to avoid unnecessary queries

Changed `from_row_with_terms` to accept a closure instead of pre-fetched term
relationships. This allows contexts to decide whether they need terms, avoiding
unnecessary database queries for contexts like Embed that don't use them.

Benefits:
- EmbedContext no longer queries term_relationships table unnecessarily
- Caller provides HOW to fetch, implementation decides IF to fetch
- Proper separation of concerns and dependency injection
- Maintains efficient batch loading for select_all (pre-fetch once, lazy access)

Implementation:
- Each context calls `fetch_terms()?` only if it needs categories/tags
- EditContext and ViewContext: Call closure to get terms
- EmbedContext: Never calls closure, avoiding the query entirely

Changes:
- Update PostContext trait method to accept FnOnce closure
- Update repository methods to pass closures instead of pre-fetching
- EditContext/ViewContext: Call fetch_terms() and use results
- EmbedContext: Don't call fetch_terms(), add explanatory comment

oguzkocer committed:
Add database migrations for posts_view_context and posts_embed_context tables

Created missing migrations for ViewContext and EmbedContext post tables to
complete the contextual repository implementation.

ViewContext table:
- Includes most fields from EditContext
- Excludes edit-only fields (password, permalink_template, generated_slug)
- Excludes raw fields (guid_raw, title_raw, content_raw, content_block_version)
- 29 columns total

EmbedContext table:
- Minimal fields for embedding posts in other content
- Only includes: id, date, link, slug, post_type, title_rendered, author,
  excerpt fields, featured_media
- 14 columns total

Changes:
- Add 0004-create-posts-view-context-table.sql migration
- Add 0005-create-posts-embed-context-table.sql migration
- Update MIGRATION_QUERIES array from 3 to 5 migrations
- Includes proper indexes and foreign key constraints
- Uses STRICT mode for type safety

oguzkocer committed:
Add integration tests with helper to verify column enum schema matching

Added PRAGMA-based tests to catch mismatches between column enum definitions
and actual database table schemas, with a reusable helper function to reduce
test duplication.

Implementation:
- Add `get_table_column_names()` helper in test_fixtures.rs
- Use helper to query actual column order from SQLite PRAGMA table_info
- Verify each enum value maps to correct column name using `.as_index()`
- Test the same code path production uses (ColumnIndex trait)
- Verify total column count matches last enum value

Benefits:
- Catches column ordering bugs immediately in CI/CD
- Reduces test duplication with shared helper function
- Self-documenting: compares enum values against expected names
- No runtime overhead: tests run during development only
- Tests fail immediately if enums get out of sync with SQL

All three context tests (Edit, View, Embed) now pass with real migrations.

oguzkocer committed:
Add upsert methods for ViewContext and EmbedContext with RETURNING optimization

Implement write operations for view and embed contexts to complete the contextual repository pattern. Use SQLite's RETURNING clause to eliminate separate SELECT queries for rowid retrieval.

Changes:
- Add `upsert()` and `upsert_batch()` methods for ViewContext and EmbedContext
- Refactor all three contexts (Edit/View/Embed) to use `RETURNING rowid` instead of separate SELECT statements
- ViewContext syncs term relationships (categories and tags)
- EmbedContext has no term relationships (minimal fields only)
- Add test to verify bundled SQLite version supports RETURNING (3.50.2 >= 3.35.0)

Performance improvements:
- Eliminates 3 database round trips per upsert (one per context)
- Safe for iOS and Android: rusqlite bundles SQLite 3.50.2 which includes RETURNING support

oguzkocer committed:
Remove unnecessary type re-exports from posts repository

The re-exports created an alternative import path without adding semantic value. Types are already well-organized in db_types::posts module where consumers can import them directly. Keeping imports in a single canonical location improves code searchability.

oguzkocer committed:
Move PostContext trait to repository and update documentation

Move entity-specific trait definitions to their repository modules for better organization. The context.rs module now contains only the generic IsContext trait and context marker types.

Changes:
- Move PostContext trait from context.rs to repository/posts.rs
- Update context.rs to focus on generic infrastructure only
- Rewrite ADDING_NEW_ENTITIES.md to be concise and accurate
- Remove outdated references to CONTEXT_FIELD_ANALYSIS.md
- Update trait method documentation to not assume all entities use term relationships
- Add reminder to update platform test migration count assertions
- Update Swift and Kotlin tests to expect 5 migrations (was 3)

Documentation improvements:
- Reduced from 487 lines to 86 lines
- Focus on general process rather than exhaustive examples
- Points to posts.rs as reference implementation
- Clarifies that entity needs determine trait method signatures

oguzkocer committed:
Remove redundant wp_mobile_cache/CLAUDE.md

The file contained outdated information referencing the old mappings/ structure and non-existent documentation files. The docs/ADDING_NEW_ENTITIES.md already serves as the canonical guide for implementing new entities.

oguzkocer committed:
Improve documentation example in get_table_column_names

Update the example to demonstrate the correct usage pattern with column
enum indices instead of hardcoded numeric indices.

Changes:
- Use `Rowid.as_index()` instead of `0`
- Use `SiteId.as_index()` instead of `1`
- Add import for `PostEditContextColumn`

jkmassel commented on wp_mobile_cache/docs/ADDING_NEW_ENTITIES.md:
```diff
@@ -0,0 +1,79 @@
+
+**Repository pattern**: All database operations belong in `repository/` - this includes both SQL execution AND row-to-type mapping logic. The `db_types/` module contains only type definitions.
+
+**RETURNING optimization**: Use SQLite's `RETURNING rowid` clause in upsert statements to eliminate separate SELECT queries. Our bundled SQLite version (3.50.2) fully supports this.
```
Probably worth mentioning the minimum iOS version supported (3.36.0), which also supports this.

oguzkocer replied:
Should I mention that the minimum supported iOS version for this library will be iOS 15 [which bundles `3.36.0`](https://github.com/yapstudios/YapDatabase/wiki/SQLite-version-(bundled-with-OS))?

jkmassel replied:
I think that's great

oguzkocer replied:
Added in 65a2f18b2d39f9a3ddc9312188cff7ee2e195782.

jkmassel commented on wp_mobile_cache/src/repository/posts.rs:
```diff
@@ -391,7 +1020,8 @@ mod tests {
             .post_repo
             .select_by_rowid(&test_ctx.conn, &test_ctx.site, rowid)
-            .unwrap();
+            .unwrap()
```
In some of these we have `unwrap().expect()` and in others we have `expect().expect()`

The latter is probably better in tests but WDYT?

oguzkocer replied:
Addressed in c1bf4f08166b7f010ded0ec89fea889d61ef4541.

jkmassel commented on wp_mobile_cache/src/lib.rs:
```diff
@@ -320,4 +324,28 @@ mod tests {
     }
+
+    #[test]
+    fn test_bundled_sqlite_version_supports_returning() {
```
I wonder if we can or should expose a function like this within the library layer, then expose it via FFI so that we can run the same check on iOS (where we use the system sqlite)

oguzkocer replied:
Added in 465f1bf5d174b23dc3b519fd8bc74d2b7ff0fe68. This is not exposed through FFI as `wp_mobile_cache` doesn't have `uniffi` as a dependency, but we can add a function to `wp_mobile` crate.

jkmassel reviewed (changes requested):
Left a few suggestions, but this generally looks solid.

oguzkocer committed:
Standardize error handling in wp_mobile_cache tests

Replace unwrap().expect() pattern with expect().expect() for better error messages.
The first expect() now handles Result errors with descriptive messages, while the
second expect() handles Option types with context-specific messages.

Changes:
- Update posts.rs tests to use expect() with "Failed to select/read post..." messages
- Update posts_constraint_tests.rs with descriptive Result error messages
- Update posts_multi_site_tests.rs for consistency
- Update posts_transaction_tests.rs for consistency

oguzkocer committed:
Extract SQLite version checking utilities for client use

Add reusable functions for querying SQLite version and checking RETURNING support.
Clients can now verify SQLite compatibility at runtime without duplicating logic.

Changes:
- Add `get_sqlite_version()` to query SQLite version via `QueryExecutor`
- Add `does_sqlite_support_returning()` to check if version >= 3.35.0
- Refactor `test_bundled_sqlite_version_supports_returning` to use new utilities
- Add unit test for version parsing with supported, unsupported, and invalid versions

jkmassel reviewed (approved):
Looks great now, thanks!!

oguzkocer committed:
Document minimum iOS 15 requirement for SQLite RETURNING support

Add platform requirements note explaining that iOS 15 (SQLite 3.36.0) is the
minimum supported version to ensure RETURNING clause compatibility.

Changes:
- Add platform requirements section in ADDING_NEW_ENTITIES.md
- Link to YapDatabase wiki for SQLite version reference by OS
